### ISSUE TITLE
"WARNING: you don't own a lock of type ExclusiveLock" in Development

### ISSUE THREAD

**Issue by newbenji** [bensheldon/good_job#388]
First of all great job. 
Working on migration some jobs to good_job.
but i get this warning out.

WARNING:  you don't own a lock of type ExclusiveLock

But i cant find documentation to support it, or im not good at finding it

**Comment by bensheldon** [comment]
@newbenji hmm, that's unexpected for you to receive that warning:

```
WARNING: you don't own a lock of type ExclusiveLock
```

The warning is output by the database adapter. It means that the current database connection (thread) is trying to unlock an Advisory Lock that does not exist. 

Can you say more about where you're seeing this?

**Comment by bensheldon** [comment]
Are you seeing this in your development environment, or in production?

Could you try adding this to an initializer (`config/initializers/good_job.rb`):

```ruby
# Output Postgres notifications through Rails.logger instead of stderr
ActiveSupport.on_load :active_record do
  ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback :checkout, :before, lambda { |conn|
    raw_connection = conn.raw_connection
    next unless raw_connection.respond_to? :set_notice_receiver

    raw_connection.set_notice_receiver do |result|
      Rails.logger.info(result.error_message)
    end
  }
end
```

...and then share the logs that are produced (the initializer above will make sure that the Postgres warning shows up in the logs, rather than going to stderr).

Also, I wrote out some of these hints too for some lock causing problems I've run across: https://github.com/bensheldon/good_job/issues/329#issuecomment-898087131

**Comment by mchristen** [comment]
I see this in my development environment! It seems to happen around the times I change code and the code reloading takes place.

**Comment by bensheldon** [comment]
Thank you @mchristen! I was able to reproduce it in development with code reloading. When a Rails reloads the code, the database connection that is used to take the initial advisory lock is not the same connection used to release the lock. I'm only seeing this when code reloading is enabled , not in a Production configuration (fhew!)

I will put together a fix for this.

**Comment by bensheldon** [comment]
I have a fix for this released in https://github.com/bensheldon/good_job/releases/tag/v2.3.1

**Comment by newbenji** [comment]
thx awsome and fast response. i must aggree i only saw the error on development

**Comment by mikehale** [comment]
I'm seeing this when running in development mode and using the external execution mode.

**Comment by bensheldon** [comment]
@mikehale hmm, that's unexpected because nothing should be taking an advisory lock with external execution mode. Could you tell me where it's coming from? Something like adding a `puts caller` in the `set_notice_receiver`

```
ActiveSupport.on_load :active_record do
  ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback :checkout, :before, lambda { |conn|
    raw_connection = conn.raw_connection
    next unless raw_connection.respond_to? :set_notice_receiver

    raw_connection.set_notice_receiver do |result|
      puts caller # <-- the new thing
    end
  }
end
```

**Comment by mikehale** [comment]
Steps to reproduce:

1. Start web and worker processes in development mode using the following initializer:

```
Rails.application.configure do
  ActiveSupport.on_load :active_record do
    ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback :checkout, :before, lambda { |conn|
      raw_connection = conn.raw_connection
      next unless raw_connection.respond_to? :set_notice_receiver

      raw_connection.set_notice_receiver do |result|
        puts caller # <-- the new thing
      end
    }
  end

  config.good_job = {
    smaller_number_is_higher_priority: true,
    execution_mode: :external,
    enable_cron: true,
    }
  }
end
```

2. Enqueue a pile of jobs to be worked
3. While the jobs are being worked modify one of my controllers
4. Notice the worker log output pause
5. Make a request to said controller
6. Wait about 30 seconds for the worker log output to resume and see the following:

```
22:30:38 worker.1 | [GoodJob] [76101] [GoodJob::Scheduler(queues=* max_threads=5)-thread-1] Executed GoodJob fb05c4c7-b848-4482-ba8b-80b99501eeaa
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:894:in `exec_params'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:894:in `block (2 levels) in exec_no_cache'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1028:in `block in with_raw_connection'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1000:in `with_raw_connection'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:893:in `block in exec_no_cache'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1143:in `log'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:892:in `exec_no_cache'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:872:in `execute_and_clear'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:64:in `internal_exec_query'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:142:in `exec_query'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/query_cache.rb:25:in `exec_query'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/concerns/good_job/advisory_lockable.rb:229:in `advisory_unlock_key'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/concerns/good_job/advisory_lockable.rb:327:in `advisory_unlock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/concerns/good_job/advisory_lockable.rb:178:in `block in with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/concerns/good_job/advisory_lockable.rb:177:in `each'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/concerns/good_job/advisory_lockable.rb:177:in `with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation/delegation.rb:79:in `block in with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:929:in `_scoping'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:467:in `scoping'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation/delegation.rb:79:in `with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/app/models/good_job/execution.rb:268:in `perform_with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation/delegation.rb:79:in `block in perform_with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:929:in `_scoping'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:467:in `scoping'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation/delegation.rb:79:in `perform_with_advisory_lock'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/lib/good_job/job_performer.rb:33:in `next'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/lib/good_job/scheduler.rb:272:in `block (2 levels) in create_task'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/reloader.rb:77:in `block in wrap'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/execution_wrapper.rb:92:in `wrap'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/reloader.rb:74:in `wrap'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/good_job-3.24.0/lib/good_job/scheduler.rb:271:in `block in create_task'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:24:in `block in execute'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `block in synchronize'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:22:in `execute'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/ivar.rb:170:in `safe_execute'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/scheduled_task.rb:298:in `process_task'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/timer_set.rb:98:in `block in ns_post_task'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:359:in `run_task'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:350:in `block (3 levels) in create_worker'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:341:in `loop'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:341:in `block (2 levels) in create_worker'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `catch'
22:30:38 worker.1 | /Users/mhale/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `block in create_worker'
```

**Comment by finchi** [comment]
Also getting this warning, in combination with `WARNING:  there is already a transaction in progress`.
But as far as I can tell, only in combination with testing on gitlab ci.

Was not able to reproduce it in development/test on my local machine or in production.

The annoying part is, that when it happens on gitlab ci, that the runner will lock up completely, until I kill the runner, or it gets killed after 1h by the timeout

It only startet to appear, after I introduced code with a lot of after_commit calls

**Comment by klevo** [comment]
I'm also getting the `WARNING:  you don't own a lock of type ExclusiveLock` in a parallelized test suite (214 tests in parallel using 12 processes). In the output It's preceded by messages like `message type 0x49 arrived from server while idle`.

It happens every 5 runs or so.

GoodJob uses stock configuration, no modifications. Rails 7.1.3.2.

<img width="666" alt="Screenshot 2024-04-04 at 14 07 14" src="https://github.com/bensheldon/good_job/assets/32981/f865aede-0c70-4b89-ba54-a6f984cd4749">

I wonder whether we should open a new issue?

**Comment by klevo** [comment]
If I disable parallelization I don't observe the above issue.

**Comment by klevo** [comment]
> Maybe open a new issue for better visibility if it is only related to test parallelization.
> 
> Briefly, I imagine that some part of GoodJob is being set up (maybe in the initial process that manages the subprocesses?) and then it's losing its database connection. That error basically means "GoodJob expected _this_ database connection to hold an advisory lock, but it doesn't" which probably means the initial database connection was closed and a new one was opened in its place.

Thanks @bensheldon . I'll see if I can create some test codebase to reliably reproduce the issue and then open an issue.
